### ISSUE TITLE
Unable to follow Quick-start: ROM Loading Failed

### ISSUE THREAD

**Issue by ssoloski** [echronos/echronos#47]
# Motivation

- Trying to follow Quick-start section to build and run an example system
- Built qemu-system-arm following steps in Prerequisites section
-- had to comment out g_mem_set_vtable() due to error: custom memory allocation vtable not supported 
- Ran qemu per Quick-start (qemu-system-arm -M simple-armv7m ...) and received error:
rom: requested regions overlap (rom phdr #3: out/machine-qemu-simple/example/rigel-system/system. free=0x000000000000161c, addr=0x00000000000013e0)
rom loading failed

# Goals

- Desired outcome - be able to run example system

# Test Plan
n/a

My system is Fedora 25; Kernel 4.10.9-200.fc25.x86_64.

**Comment by stefangotz** [comment]
Hi Steve,

thanks for reporting this!

I actually have the branch https://github.com/echronos/echronos/tree/simplify_getting_started_guide in the queue for improving the _quick-start_ guide, but it will probably take a little until it is ready to go.

Meanwhile, can we help you with getting started? Are you specifically interested in an ARM or PowerPC based sample system?

**Comment by ssoloski** [comment]
Hi Stefan,
Thanks for the reply - I would love some help getting this up and running! I'm specifically interested in the ARM system. We're running a fork of the (almost) latest version of QEMU; once I get this running under your QEMU version I'll probably end up porting code into our fork for our testing. But I would be more than happy to help with anything you need to get this running under ARM - let me know what I can do to help.

Do you want me to take a look at (and try?) the guide in your branch?

Steve

**Comment by stefangotz** [comment]
Hi Steve,

I talked to Se in our team and he'll be in touch over the coming days to help out. (He is much more across the ARM port than I am).

Regards,
     Stefan

**Comment by ssoloski** [comment]
Hey Stefan,
I talked to my manager, and there's no hurry on this - turns out he wants to me to look at seL4 instead. So I'm off to check it out - but if you want me to try anything let me know.
Steve

**Comment by stefangotz** [comment]
Hi Steve,

no worries - definitely do get in touch with the seL4 team if you have any questions. I work with them from time to time and they're a super helpful bunch!

Stefan

**Comment by schnommus** [comment]
Just wanted to follow this up in case someone else runs into this issue and reads this.

To get QEMU working on ARM, there are a couple things you need to do in addition to the current README contents. If you are using a modern version of GCC, when you compile the Breakaway QEMU fork you'll need to additionally supply the `--disable-gtk` option, so your config line will look like:
```
./configure --target-list=arm-softmmu --disable-cocoa --disable-curses --disable-vnc --disable-tools --without-pixman --disable-console --disable-default-devices --disable-slirp --disable-curl --disable-guest-base --disable-guest-agent --disable-blobs --audio-drv-list= --audio-card-list= --disable-usb --disable-ide --disable-pie --enable-debug --disable-sdl --disable-fdt --disable-spice --disable-xen --disable-werror --disable-gtk
```
Additionally, you will have to modify the QEMU source code to bypass a known issue with NOLOAD sections (report [here](https://bugs.launchpad.net/qemu/+bug/1429841))
```
--- a/hw/loader.c
+++ b/hw/loader.c
@@ -721,7 +721,7 @@ int rom_load_all(void)
                     "(rom %s. free=0x" TARGET_FMT_plx
                     ", addr=0x" TARGET_FMT_plx ")\n",
                     rom->name, addr, rom->addr);
-            return -1;
+            //return -1;
         }
         addr  = rom->addr;
         addr += rom->romsize;
```
You can then `make` QEMU and run the example system.

**Comment by stefangotz** [comment]
Awesome, thanks Sebastian! I will address these issues in the README as time permits.

**Comment by stefangotz** [comment]
Created https://github.com/echronos/echronos/tree/47_unable_to_follow_quick-start_rom_loading_failed to resolve this issue.

**Comment by ssoloski** [comment]
First off - thanks for the help guys!

But... even with those changes it still doesn't seem to work.

I start up QEMU, and get the error about requested regions overlap - not a problem as the change still lets the error message print out.
In a second window, I fire up gdb, and here's the output:

Reading symbols from out/machine-qemu-simple/example/rigel-system/system...done.
Remote debugging using :1234
entry () at out/machine-qemu-simple/example/rigel-system/armv7m/vectable.s:321
321	        ldr r0, =data_load_addr
Breakpoint 1 at 0xffc: file packages/rtos-example/rigel-test.c, line 50.
(gdb) c
Continuing.

Breakpoint 1, fatal (error_id=error_id@entry=1 '\001') at packages/rtos-example/rigel-test.c:50
50	{
(gdb) 

According to the quick-start I should be seeing:

task a: start delay
irq tick
irq tick
irq tick
irq tick
task a: complete delay

Am I missing something else?
Steve

**Comment by schnommus** [comment]
Hi Steve,

Thanks for testing out the changes! I should have kept this thread in the loop - I can confirm your problem as I also came across the issue you state when I was working on https://github.com/echronos/echronos/tree/47_unable_to_follow_quick-start_rom_loading_failed .

It has been fixed in that branch but is currently awaiting review before it gets integrated. Essentially the issue is that semihosting debug output only appears when gdb is not attached to the machine. You'll find that if you remove the `-S` option on the qemu line that the machine will work as desired (albeit without gdb). When I initially made that post before I'd tested the machine by itself as I (mistakedly) didn't expect it to behave differently when gdb was attached.

I found some archaic QEMU bug reports that talked about this issue, so the solution in the new branch is to use a completely different (and new) version of QEMU: https://github.com/echronos/qemu

Feel free to test out the fixes in the `47_unable_to_follow_quick-start_rom_loading_failed` branch (i.e the readme changes), or wait a bit and they should be integrated in the near future. I intend on adding the quickstart instructions to our regression suite in the future so issues like this won't happen again.

Also I would like to note that our ARM support has been tested on hardware platforms, it's just that the QEMU support seems to have rotted a bit. This will be fixed very soon!

Cheers,
Seb.

**Comment by ssoloski** [comment]
Thanks for the info, Seb - I'll try out the new QEMU and the new branch today.

A bit off topic, but I was looking for (and didn't find) something talking about which hardware platforms are supported for eChronos. I'm trying to run on Pi 2 (or 3), and then potentially on a BeagleBone Black device. I initially was looking at seL4, but it doesn't support any of those devices. So now I'm back to looking at eChronos... do you guys support any of those platforms?

Thanks again for all the help!

Steve

**Comment by schnommus** [comment]
Hi Steve,

It would be theoretically possible for you to run posix-compatible eChronos systems on a beaglebone or raspberry pi linux system; however unfortunately we don't yet have native support for the ARMv7 (bbb) or ARMv8A (rpi3) architectures in the eChronos RTOS, only ARMv7m (cortex-m processors) - different instruction set. These larger processors with MMUs are a bit more powerful than the sorts of processors that eChronos is currently purposed for.

Just an FYI, seL4 does actually support the beaglebone black and raspberry pi 3 according to their hardware support page: https://wiki.sel4.systems/Hardware. If you're running into issues with that platform support I encourage you to get in contact with the seL4 developers and let them know as bug reports are always useful :).

Cheers,
Seb.

**Comment by ssoloski** [comment]
Thanks for the info, Seb! I tried out the branch build and had no problems with it, other than the QEMU setup... personally, I would prefer to do a make install rather than have to set the path to ./path. But that's just me - bottom line is the code worked fine.

I'll double check the seL4 page again; I didn't see those systems listed but I might have been looking at an older page. 

But thanks again for all your help and info - really nice to get help like this!

Steve

**Comment by schnommus** [comment]
Fixed in v1.3.1. task: 47_unable_to_follow_quick-start_rom_loading_failed.
